### PR TITLE
MM-59935: Add system console UI for Redis

### PR DESCRIPTION
**Pull request by agnivade**
https://mattermost.atlassian.net/browse/MM-59935

```release-note
Redis is now available as an alternative cache backend for all enterprise customers. It can be leveraged to run Mattermost at very high scale. There is a new section in config.json that can be updated to use Redis.

CacheSettings.CacheType: This can be either lru or redis. "lru" is the default choice which will use the in-memory cache
store that we use currently.
CacheSettings.RedisAddress: The hostname of the Redis host
CacheSettings.RedisPassword: The password of the Redis host. (Can be left blank if there is no password)
CacheSettings.RedisDB: The database of the Redis host. Typically 0.
CacheSettings.DisableClientCache: This can be set to true if you decide to disable the client-side cache of Redis. Typically there is no need to do this in production, and this is mainly used as a test option.
```
;

### CODE REVIEW

**Comment by agnivade** [comment]
Blocked on https://github.com/mattermost/mattermost/pull/28917

File: webapp/channels/src/components/admin_console/admin_definition.tsx [hunk 1]
```diff
@@ -294,6 +294,24 @@ const getRestrictedIndicator = (displayBlocked = false, minimumPlanRequiredForFe
...
+
+    redis_address_title: {id: 'admin.cacheSettings.redisAddress', defaultMessage: 'Redis Address'},
+    redis_address_desc: {id: 'admin.cacheSettings.redisAddressDesc', defaultMessage: 'The hostname:port of the Redis server. E.g.: "localhost:6379"'},
+    redis_address_placeholder: {id: 'admin.cacheSettings.redisAddressPlaceholder', defaultMessage: 'localhost:6379'},
+
+    redis_password_title: {id: 'admin.cacheSettings.redisPassword', defaultMessage: 'Redis Password'},
+    redis_password_desc: {id: 'admin.cacheSettings.redisPasswordDesc', defaultMessage: 'The password of the Redis server.'},
+
+    redis_db_title: {id: 'admin.cacheSettings.redisDB', defaultMessage: 'Redis DB'},
+    redis_db_desc: {id: 'admin.cacheSettings.redisDBDesc', defaultMessage: 'The database of the Redis server. E.g.: "0"'},
```
Comment (agarciamontoro) [inline webapp/channels/src/components/admin_console/admin_definition.tsx:310]: What does `0` mean here?

**Review by agarciamontoro — APPROVED** [review]
Code looks good! Should we ask our technical writer for a review to make all sentences consistent? I see there are differences in casing for example.

**Review by cpoile — APPROVED** [review]
It's complicated and annoying to add a new panel, nice work. :)

File: webapp/channels/src/components/admin_console/admin_definition.tsx (hunk 1 above)
Comment (cpoile) [reply webapp/channels/src/components/admin_console/admin_definition.tsx:310]: Redis databases are identified by an integer index.

Comment (agnivade) [reply webapp/channels/src/components/admin_console/admin_definition.tsx:310]: What Chris said :point_up_2:

**Comment by agnivade** [comment]
> It's complicated and annoying to add a new panel, nice work. :)

Actually, I found the process more or less manageable. More so because there wasn't a need to create a new component with yet more boilerplate code. The only bother were the translation strings, but it was alright.

**Comment by agnivade** [comment]
@DHaussermann - Please test this after https://github.com/mattermost/mattermost/pull/28917.

Steps:
- Verify that the system console section appears correctly.
- Verify that it's not shown if there's no enterprise license.
- Verify you can save the config changes. (Although you'll need to restart MM for changes to take place)

**Comment by agnivade** [comment]
@cwarnermm - Would you mind looking at the translation strings and verify if anything needs changing? Thanks!

File: webapp/channels/src/i18n/en.json [hunk 2]
```diff
@@ -431,6 +431,21 @@
...
   "admin.bleve.title": "Bleve",
+  "admin.cacheSettings.cacheType.lru": "LRU",
+  "admin.cacheSettings.cacheType.redis": "Redis",
+  "admin.cacheSettings.cacheTypeDesc": "The type of the cache backend. E.g.: \"redis\" or \"lru\"",
+  "admin.cacheSettings.cacheTypeTitle": "Cache Type",
+  "admin.cacheSettings.redisAddress": "Redis Address",
+  "admin.cacheSettings.redisAddressDesc": "The hostname:port of the Redis server. E.g.: \"localhost:6379\"",
+  "admin.cacheSettings.redisAddressPlaceholder": "localhost:6379",
+  "admin.cacheSettings.redisClientCache": "Disable Client Cache",
+  "admin.cacheSettings.redisClientCacheDesc": "When true, client side caching is disabled.",
```
Comment (cwarnermm) [inline webapp/channels/src/i18n/en.json:442]:
```suggestion
  "admin.cacheSettings.redisClientCacheDesc": "When true, client-side caching is disabled.",
```

File: webapp/channels/src/components/admin_console/admin_definition.tsx [hunk 3]
```diff
@@ -294,6 +294,24 @@ const getRestrictedIndicator = (displayBlocked = false, minimumPlanRequiredForFe
...
+    redis_db_placeholder: {id: 'admin.cacheSettings.redisDBPlaceholder', defaultMessage: '0'},
+
+    redis_clientcache_title: {id: 'admin.cacheSettings.redisClientCache', defaultMessage: 'Disable Client Cache'},
+    redis_clientcache_desc: {id: 'admin.cacheSettings.redisClientCacheDesc', defaultMessage: 'When true, client side caching is disabled.'},
```
Comment (cwarnermm) [inline webapp/channels/src/components/admin_console/admin_definition.tsx:314]:
```suggestion
    redis_clientcache_desc: {id: 'admin.cacheSettings.redisClientCacheDesc', defaultMessage: 'When true, client-side caching is disabled.'},
```

**Review by cwarnermm — APPROVED** [review]
Provided editorial feedback inline. Approving.

**Comment by DHaussermann** [comment]
@agnivade For some reason, I see something strange on my local setup.
No issue seeing the UI and having the UI disappear based on license. But I don't seem to be able to use the UI to update my config.

When I save changes via the UI and restart my server the values go back to default and the config is still unchanged.

For fun, I also tried updating the values in my config file directly and saving then restarting my server but the UI did not show these changes either. Could this be possible or do I have some type of setup issue?
  
Please advise if you have any thoughts on isolation steps. I did not catch if there was anything relevant in the server logs to clarify whats happening.

**Comment by agnivade** [comment]
Are you able to save other config settings, except the Redis ones?

Happy to jump on a quick call later today.

**Comment by agnivade** [comment]
Also, priority wise, it would be good to review https://github.com/mattermost/mattermost/pull/28917 first, and then come to this. But the issues you are seeing are unrelated. Anyways, LMK.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- System console is shown as expected
- Cache config page is removed for Professional or Team addition
- Verified changing config manually get's reflected in new UI and that saving updates to the UI will update the config

LGTM!

**Comment by agnivade** [comment]
@cwarnermm - Please feel free to update the release documentation for when this goes out. Now we can openly announce Redis support. 

cc @amyblais

**Pull request merged**